File: ortools/constraint_solver/samples/nurses_cp.cc
// Copyright 2018 Google LLC
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <numeric>  // std::iota

#include "ortools/base/logging.h"
#include "ortools/constraint_solver/constraint_solver.h"

namespace operations_research {
void SolveNursesExample() {
  // Instantiate the solver.
  Solver solver("NursesExample");
  std::array<int, 4> nurses;
  std::iota(std::begin(nurses), std::end(nurses), 0);
  {
    std::ostringstream oss;
    for (auto i : nurses) oss << ' ' << i;
    LOG(INFO) << "Nurses:" << oss.str();
  }

  // Nurse assigned to shift 0 means not working that day.
  std::array<int, 4> shifts;
  std::iota(std::begin(shifts), std::end(shifts), 0);
  {
    std::ostringstream oss;
    for (auto i : shifts) oss << ' ' << i;
    LOG(INFO) << "Shifts:" << oss.str();
  }

  std::array<int, 7> days;
  std::iota(std::begin(days), std::end(days), 0);
  {
    std::ostringstream oss;
    for (auto i : days) oss << ' ' << i;
    LOG(INFO) << "Days:" << oss.str();
  }

  // Create shift variables.
  std::vector<std::vector<IntVar*>> shifts_matrix(nurses.size());
  std::vector<IntVar*> shifts_flat;
  for (const auto nurse : nurses) {
    for (const auto day : days) {
      std::ostringstream oss;
      oss << "shifts(nurse: " << nurse << ", day: " << day << ")";
      IntVar* var = solver.MakeIntVar(shifts.front(), shifts.back(), oss.str());
      shifts_matrix[nurse].push_back(var);
      shifts_flat.push_back(var);
    }
  }

  // Create nurse variables.
  std::vector<std::vector<IntVar*>> nurses_matrix(shifts.size());
  for (const auto shift : shifts) {
    for (const auto day : days) {
      std::ostringstream oss;
      oss << "nurses(shift: " << shift << ", day: " << day << ")";
      IntVar* var = solver.MakeIntVar(nurses.front(), nurses.back(), oss.str());
      nurses_matrix[shift].push_back(var);
    }
  }

  // Set relationships between shifts and nurses.
  for (const auto day : days) {
    std::vector<IntVar*> nurses_for_day(shifts.size());
    for (const auto shift : shifts) {
      nurses_for_day[shift] = nurses_matrix[shift][day];
    }
    for (const auto nurse : nurses) {
      IntVar* s = shifts_matrix[nurse][day];
      solver.AddConstraint(
          solver.MakeEquality(solver.MakeElement(nurses_for_day, s), nurse));
    }
  }

  // Make assignments different on each day i.e.
  for (const auto day : days) {
    // no shift can have two nurses
    std::vector<IntVar*> shifts_for_day(nurses.size());
    for (const auto nurse : nurses) {
      shifts_for_day[nurse] = shifts_matrix[nurse][day];
    }
    solver.AddConstraint(solver.MakeAllDifferent(shifts_for_day));

    // no nurses can have more than one shifts a day
    std::vector<IntVar*> nurses_for_day(shifts.size());
    for (const auto shift : shifts) {
      nurses_for_day[shift] = nurses_matrix[shift][day];
    }
    solver.AddConstraint(solver.MakeAllDifferent(nurses_for_day));
  }

  // Each nurse works 5 or 6 days in a week.
  for (const auto nurse : nurses) {
    std::vector<IntVar*> nurse_is_working;
    for (const auto day : days) {
      nurse_is_working.push_back(
          solver.MakeIsGreaterOrEqualCstVar(shifts_matrix[nurse][day], 1));
    }
    solver.AddConstraint(solver.MakeSumGreaterOrEqual(nurse_is_working, 5));
    solver.AddConstraint(solver.MakeSumLessOrEqual(nurse_is_working, 6));
  }

  // Create works_shift variables.
  // works_shift_matrix[n][s] is True if
  // nurse n works shift s at least once during the week.
  std::vector<std::vector<IntVar*>> works_shift_matrix(nurses.size());
  for (const auto nurse : nurses) {
    for (const auto shift : shifts) {
      std::ostringstream oss;
      oss << "work_shift(nurse: " << nurse << ", shift: " << shift << ")";
      works_shift_matrix[nurse].push_back(solver.MakeBoolVar(oss.str()));
    }
  }

  for (const auto nurse : nurses) {
    for (const auto shift : shifts) {
      std::vector<IntVar*> shift_s_for_nurse;
      for (const auto day : days) {
        shift_s_for_nurse.push_back(
            solver.MakeIsEqualCstVar(shifts_matrix[nurse][day], shift));
      }
      solver.AddConstraint(
          solver.MakeEquality(works_shift_matrix[nurse][shift],
                              solver.MakeMax(shift_s_for_nurse)->Var()));
    }
  }

  // For each shift(other than 0), at most 2 nurses are assigned to that shift
  // during the week.
  for (std::size_t shift = 1; shift < shifts.size(); ++shift) {
    std::vector<IntVar*> nurses_for_shift;
    for (const auto nurse : nurses) {
      nurses_for_shift.push_back(works_shift_matrix[nurse][shift]);
    }
    solver.AddConstraint(solver.MakeSumLessOrEqual(nurses_for_shift, 2));
  }

  // If a nurse works shifts 2 or 3 on,
  // he must also work that shift the previous day or the following day.
  for (const auto shift : {2, 3}) {
    for (const auto day : days) {
      IntVar* v0 = solver.MakeIsEqualVar(nurses_matrix[shift][day],
                                         nurses_matrix[shift][(day + 1) % 7]);
      IntVar* v1 = solver.MakeIsEqualVar(nurses_matrix[shift][(day + 1) % 7],
                                         nurses_matrix[shift][(day + 2) % 7]);
      solver.AddConstraint(solver.MakeEquality(solver.MakeMax(v0, v1), 1));
    }
  }

  // ----- Search monitors and decision builder -----

  // Create the decision builder.
  DecisionBuilder* const main_phase = solver.MakePhase(
      shifts_flat, Solver::CHOOSE_FIRST_UNBOUND, Solver::ASSIGN_MIN_VALUE);

  // Search log.
  SearchMonitor* const search_log = nullptr;

  SearchLimit* limit = nullptr;

  // Create the solution collector.
  SolutionCollector* const collector = solver.MakeAllSolutionCollector();
  collector->Add(shifts_flat);

  // Solve
  solver.Solve(main_phase, search_log, nullptr, limit, collector);
  LOG(INFO) << "Number of solutions: " << collector->solution_count();
  LOG(INFO) << "";

  // Display a few solutions picked at random.
  std::array<int, 4> a_few_solutions = {859, 2034, 5091, 7003};
  for (const auto solution : a_few_solutions) {
    LOG(INFO) << "Solution " << solution << ":";
    for (const auto day : days) {
      LOG(INFO) << "Day " << day << ":";
      for (const auto nurse : nurses) {
        LOG(INFO) << "Nurse " << nurse << " assigned to "
                  << "Task "
                  << collector->Value(solution,
                                      shifts_flat[nurse * days.size() + day]);
      }
    }
  }
  LOG(INFO) << "Advanced usage:";
  LOG(INFO) << "Time: " << solver.wall_time() << "ms";
}
}  // namespace operations_research

int main(int argc, char** argv) {
  google::InitGoogleLogging(argv[0]);
  absl::SetFlag(&FLAGS_logtostderr, 1);
  operations_research::SolveNursesExample();
  return EXIT_SUCCESS;
}

File: examples/contrib/nurses_cp.py
import sys

from ortools.constraint_solver.python import constraint_solver as cp


def main():
    # Creates the solver.
    solver = cp.Solver("schedule_shifts")

    num_nurses = 4
    num_shifts = 4  # Nurse assigned to shift 0 means not working that day.
    num_days = 7
    # [START]
    # Create shift variables.
    shifts = {}

    for j in range(num_nurses):
        for i in range(num_days):
            shifts[(j, i)] = solver.new_int_var(
                0, num_shifts - 1, "shifts(%i,%i)" % (j, i)
            )
    shifts_flat = [shifts[(j, i)] for j in range(num_nurses) for i in range(num_days)]

    # Create nurse variables.
    nurses = {}

    for j in range(num_shifts):
        for i in range(num_days):
            nurses[(j, i)] = solver.new_int_var(
                0, num_nurses - 1, "shift%d day%d" % (j, i)
            )
    # Set relationships between shifts and nurses.
    for day in range(num_days):
        nurses_for_day = [nurses[(j, day)] for j in range(num_shifts)]

        for j in range(num_nurses):
            s = shifts[(j, day)]
            solver.add(s.index_of(nurses_for_day) == j)
    # Make assignments different on each day
    for i in range(num_days):
        solver.add_all_different([shifts[(j, i)] for j in range(num_nurses)])
        solver.add_all_different([nurses[(j, i)] for j in range(num_shifts)])
    # Each nurse works 5 or 6 days in a week.
    for j in range(num_nurses):
        solver.add(solver.sum([shifts[(j, i)] > 0 for i in range(num_days)]) >= 5)
        solver.add(solver.sum([shifts[(j, i)] > 0 for i in range(num_days)]) <= 6)
    # Create works_shift variables. works_shift[(i, j)] is True if nurse
    # i works shift j at least once during the week.
    works_shift = {}

    for i in range(num_nurses):
        for j in range(num_shifts):
            works_shift[(i, j)] = solver.new_bool_var("nurse%d shift%d" % (i, j))

    for i in range(num_nurses):
        for j in range(num_shifts):
            solver.add(
                works_shift[(i, j)]
                == solver.max([shifts[(i, k)] == j for k in range(num_days)])
            )

    # For each shift (other than 0), at most 2 nurses are assigned to that shift
    # during the week.
    for j in range(1, num_shifts):
        solver.add(solver.sum([works_shift[(i, j)] for i in range(num_nurses)]) <= 2)
    # If s nurses works shifts 2 or 3 on, he must also work that shift the previous
    # day or the following day.
    solver.add(
        solver.max(nurses[(2, 0)] == nurses[(2, 1)], nurses[(2, 1)] == nurses[(2, 2)])
        == 1
    )
    solver.add(
        solver.max(nurses[(2, 1)] == nurses[(2, 2)], nurses[(2, 2)] == nurses[(2, 3)])
        == 1
    )
    solver.add(
        solver.max(nurses[(2, 2)] == nurses[(2, 3)], nurses[(2, 3)] == nurses[(2, 4)])
        == 1
    )
    solver.add(
        solver.max(nurses[(2, 3)] == nurses[(2, 4)], nurses[(2, 4)] == nurses[(2, 5)])
        == 1
    )
    solver.add(
        solver.max(nurses[(2, 4)] == nurses[(2, 5)], nurses[(2, 5)] == nurses[(2, 6)])
        == 1
    )
    solver.add(
        solver.max(nurses[(2, 5)] == nurses[(2, 6)], nurses[(2, 6)] == nurses[(2, 0)])
        == 1
    )
    solver.add(
        solver.max(nurses[(2, 6)] == nurses[(2, 0)], nurses[(2, 0)] == nurses[(2, 1)])
        == 1
    )

    solver.add(
        solver.max(nurses[(3, 0)] == nurses[(3, 1)], nurses[(3, 1)] == nurses[(3, 2)])
        == 1
    )
    solver.add(
        solver.max(nurses[(3, 1)] == nurses[(3, 2)], nurses[(3, 2)] == nurses[(3, 3)])
        == 1
    )
    solver.add(
        solver.max(nurses[(3, 2)] == nurses[(3, 3)], nurses[(3, 3)] == nurses[(3, 4)])
        == 1
    )
    solver.add(
        solver.max(nurses[(3, 3)] == nurses[(3, 4)], nurses[(3, 4)] == nurses[(3, 5)])
        == 1
    )
    solver.add(
        solver.max(nurses[(3, 4)] == nurses[(3, 5)], nurses[(3, 5)] == nurses[(3, 6)])
        == 1
    )
    solver.add(
        solver.max(nurses[(3, 5)] == nurses[(3, 6)], nurses[(3, 6)] == nurses[(3, 0)])
        == 1
    )
    solver.add(
        solver.max(nurses[(3, 6)] == nurses[(3, 0)], nurses[(3, 0)] == nurses[(3, 1)])
        == 1
    )
    # Create the decision builder.
    db = solver.phase(
        shifts_flat,
        cp.IntVarStrategy.CHOOSE_FIRST_UNBOUND,
        cp.IntValueStrategy.ASSIGN_MIN_VALUE,
    )
    # Create the solution collector.
    solution = solver.assignment()
    solution.add(shifts_flat)
    collector = solver.all_solution_collector(solution)

    solver.solve(db, [collector])
    print("Solutions found:", collector.solution_count)
    print("Time:", solver.wall_time_ms, "ms")
    print()
    # Display a few solutions picked at random.
    a_few_solutions = [859, 2034, 5091, 7003]

    for sol in a_few_solutions:
        print("Solution number", sol, "\n")

        for i in range(num_days):
            print("Day", i)
            for j in range(num_nurses):
                print(
                    "Nurse", j, "assigned to task", collector.value(sol, shifts[(j, i)])
                )
            print()


if __name__ == "__main__":
    main()
